### ISSUE TITLE
PyTools

### ISSUE THREAD

**Issue by DasMoorhuhn** [hax0r31337/denpk2#4]
Ay^^

From where can I get the pytools lib which is linked in the `Cargo.toml`? `pytools = { path = "../pytools" }`

**Comment by KKeygen** [comment]
I'm currently working on remapping opcodes using a modified version of the `xdis` library, which I thought may work as an alternative of `pytools`. I got the opcode mapping from the latest commit. I got pyc files from the last commit of this repo without `pytools` dependency. I fixed the magic number of the `.pyc` file and have the opcode remapped as shown below:

```python
(
    filename,
    co,
    version,
    timestamp,
    magic_int,
    is_pypy,
    source_size,
    sip_hash,
) = xdis.disassemble_file(
    str(pyc_file), outstream=open(os.devnull, "w"), alternate_opmap=alternate_opmap #from src/opcode/mapping.rs
)
```

However, the output `pyc` file can't be decompiled. I have tried using both `uncompyle6` and `pycdc`, and they both warned about an unsupported opcode. For `pycdc`, the specific unsupported opcode was `MAKE_CELL`.

I suspect there might be an issue with how the opcode is being remapped. To debug this further, the access to the `pytools` dependency should be helpful.

@hax0r31337 
Could you please make the `pytools`  public? This will greatly help in locating and fixing the error with decompiling, thanks!

**Comment by hax0r31337** [comment]
I was planning to add decompiler to `pytools` library, but I having trouble analyzing dominator tree of python bytecode.    
That's the reason I didn't made `pytools` available publicly.    

btw both `pycdc` and `uncompyle6` unable to decompile Python 3.11 bytecode, although `pycdc` claims the capability

**Comment by n1tr0ms** [comment]
Did the fix will be someday?

**Comment by wuqu223** [comment]
> 我原本计划将反编译器添加到`pytools`库中，但在分析 Python 字节码的支配树时遇到了麻烦。 这就是我没有`pytools`公开的原因。
> 
> 顺便说一下，两者都`pycdc`无法`uncompyle6`反编译 Python 3.11 字节码，尽管`pycdc`声称有能力

Hello, you are keen on studying the decompilation work of Identity V. Can you upload an old version of the opcode map? I will try to use this mapping to redirect the pyc file to recommend the opcodes of other NetEase games (still using python2.7) Mapping

**Comment by hax0r31337** [comment]
> other NetEase games (still using python2.7)

Every different game on neox/messiah engine has it's own set of opcodes, they prob automated the opcode randomization process.   
We have to thank them for not updating opcode every version, but even if they do, opcode can be tracked by comparing versions.

**Comment by wuqu223** [comment]
> > 其他网易游戏（仍使用python2.7）
> 
> neox/messiah 引擎上的每个不同游戏都有自己的一组操作码，它们可能自动执行了操作码随机化过程。 我们必须感谢他们没有在每个版本都更新操作码，但即使他们这样做，也可以通过比较版本来跟踪操作码。

Yes, I used a redirect script for a certain game to map several other NetEase games.
I've completed opcode mapping for several games for my personal work, and they may become invalid as the engine is upgraded
This should be the easiest way to get the opcode, although there are some problems, because I don't work in reverse, some so files seem to be passive
I am worried that the Python version of a certain game will be upgraded too high in the future, resulting in no suitable decompilation tool processing

**Comment by hax0r31337** [comment]
> > > 其他网易游戏（仍使用python2.7）
> > 
> > 
> > neox/messiah 引擎上的每个不同游戏都有自己的一组操作码，它们可能自动执行了操作码随机化过程。 我们必须感谢他们没有在每个版本都更新操作码，但即使他们这样做，也可以通过比较版本来跟踪操作码。
> 
> Yes, I used a redirect script for a certain game to map several other NetEase games. I've completed opcode mapping for several games for my personal work, and they may become invalid as the engine is upgraded This should be the easiest way to get the opcode, although there are some problems, because I don't work in reverse, some so files seem to be passive I am worried that the Python version of a certain game will be upgraded too high in the future, resulting in no suitable decompilation tool processing

I'm not interested in NetEase games rn, I'm busy making my private CS2 HvH cheat.

**Comment by wuqu223** [comment]
> > 其他网易游戏（仍使用python2.7）
> 
> neox/messiah 引擎上的每个不同游戏都有自己的一组操作码，它们可能自动执行了操作码随机化过程。 我们必须感谢他们没有在每个版本都更新操作码，但即使他们这样做，也可以通过比较版本来跟踪操作码。

Can you reverse engineering? Identity V unknown opcodes are too many, I can't compare the work, I can only compare some games that use fewer custom opcodes

**Comment by hax0r31337** [comment]
> > > 其他网易游戏（仍使用python2.7）
> > 
> > 
> > neox/messiah 引擎上的每个不同游戏都有自己的一组操作码，它们可能自动执行了操作码随机化过程。 我们必须感谢他们没有在每个版本都更新操作码，但即使他们这样做，也可以通过比较版本来跟踪操作码。
> 
> Can you reverse engineering? Identity V unknown opcodes are too many, I can't compare the work, I can only compare some games that use fewer custom opcodes

Yes I can, but I won't, and **learn how to speak politely when asking for a favor**

**Comment by wuqu223** [comment]
Have you encountered it
Unsupported opcodes (MAP_ADD) I think this may be a key issue with not being able to decompile text py

**Comment by compilelinking-wq** [comment]
> 您是否遇到过 不支持的操作码（MAP_ADD）我认为这可能是无法反编译文本 py 的一个关键问题
This is a problem with the decompiler, not with the opcode

**Comment by compilelinking-wq** [comment]
> Have you encountered it Unsupported opcodes (MAP_ADD) I think this may be a key issue with not being able to decompile text py

This is a problem with the decompiler, not with the opcode

**Comment by G3NuitR** [comment]
> was planning to add decompiler to `pytools` library, but I having trouble analyzing dominator tree of python bytecode.

Hi, thank you very much for making this repository public, it has given me great inspiration.

Recently, I am working on decrypting code files from another game running on NeoX2. I have located its decrypt method. These mostly inspired by this repo.

But when I want to try to add the decompressing logic into the code and run it, i found the same problem mentioned in this issue, the `pytools` crate is missing. Could you please make it public or send me a copy? That would be very helpful!

**Comment by hax0r31337** [comment]
> > was planning to add decompiler to `pytools` library, but I having trouble analyzing dominator tree of python bytecode.
> 
> Hi, thank you very much for making this repository public, it has given me great inspiration.
> 
> Recently, I am working on decrypting code files from another game running on NeoX2. I have located its decrypt method. These mostly inspired by this repo.
> 
> But when I want to try to add the decompressing logic into the code and run it, i found the same problem mentioned in this issue, the `pytools` crate is missing. Could you please make it public or send me a copy? That would be very helpful!

I'm not planning to release it, as the library is underdeveloped and still lacks some essential feature.   
I was originally trying to design it's API with [Java ObjectWeb ASM](https://asm.ow2.io/) style in mind but lose interest and work on another project.   

But anyway it's possible to implement one by reading Python interpreter code in Marshal format parsing within hours.

**Comment by G3NuitR** [comment]
> I'm not planning to release it, as the library is underdeveloped and still lacks some essential feature. I was originally trying to design it's API with [Java ObjectWeb ASM](https://asm.ow2.io/) style in mind but lose interest and work on another project.
> 
> But anyway it's possible to implement one by reading Python interpreter code in Marshal format parsing within hours.

Thanks for reply! It is completely understandable.
